spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindHidingSubClass.java

```diff
@@ -0,0 +1,146 @@
+                if (method.isStatic() && !method.isPrivate() && !isMainMethod(method)) {
+                    //I check for the exceptional cases of inner class using three auxiliary private methods.
+                    if (isConstructor(method) || isHidingInnerClass(method) || isAutoGeneratedMethod(method)) {
+                        continue;
+                    }
+                    //I check either the subclass method is hiding the superclass method. If yes I report the bug.
+                    if (Arrays.asList(superMethods).contains(method)) {
+                        bugReporter.reportBug(new BugInstance(this, "HSBC_HIDING_SUB_CLASS", NORMAL_PRIORITY)
+                                .addClass(subClass.getClassName())
+                                .addMethod(subClass, method)
+                                .addClass(superClass.getClassName())
+                                .addSourceLine(SourceLineAnnotation.fromVisitedInstruction(subClass, method, 0)));
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void report() {
+    }
+
+    /**
+     *This method checks for the inner class exceptional cases.
+     *As whenever there an inner class, '.access$' methods are created hiddenly to access the outer class attributes.
+     */
+    private boolean isHidingInnerClass(Method method) {
+        return method.getName().contains("access$");
+    }
+
+    /**
+     *This method checks for the autoGenerated methods for an inner class - exceptional cases.
+     *As whenever there an inner class, '.class$' methods are created hiddenly.
+     */
+    private boolean isAutoGeneratedMethod(Method method) {
+        return method.getName().contains("class$");
+    }
+
+    /**
+     * This method is here to check the exceptional case of Constructors
+     */
+    private boolean isConstructor(Method method) {
+        return Const.STATIC_INITIALIZER_NAME.equals(method.getName()) || Const.CONSTRUCTOR_NAME.equals(method.getName());
+    }
+
+    /**
+     * This method checks for the exceptional case of main method.
+     * As we know main method always have the signature "public static void main(String[] args)".
+     * It is static but usually public class have its own main method as its entry point.
+     * Therefore, it is not an error caused by Programming but a utility to provide UI to user.
+     *
+     * Frist part of conditional is to abide with the  @see <a href="https://openjdk.org/jeps/445">JEP 445</a>
+     *
+     */
+    private boolean isMainMethod(Method method) {
+        return (method.getReturnType().equals(BasicType.VOID) && "main".equals(method.getName())) ||
+                (method.isPublic() && method.isStatic() && method.getReturnType().equals(BasicType.VOID)
+                && "main".equals(method.getName())
+                && isStringArray(method.getArgumentTypes()));
+    }
```

Comment:
In the current form, the second part of the condition is unnecessary, since the second half repeats the first half, it's always false when evaluated. Can you please modify the condition to not contain any unnecessary parts, but filter out the "mainlike" functions. The "Selecting a main method" part in https://openjdk.org/jeps/445 could be really useful.

Reply:
I have updated the condition.